test/reset/main.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+ * PackageLicenseDeclared: Apache-2.0
+ * Copyright (c) 2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* @file main.cpp
+ *
+ * This test suite tests scenarios to reset target in-between the test.
+ *
+ *   The following tests are implemented in this file:
+ *     Test 1            Reset target inbetween the tests.
+ */
+
+#include "greentea-client/test_env.h"
+#include "utest/utest.h"
+#include "unity/unity.h"
+
+#include "mbed-drivers/mbed.h"
+
+#include <algorithm>
+#include <cstdlib>
+
+using namespace utest::v1;
+
+/** Custom echo test
+ * 
+ *  This function sends key,value pairs like {{echo;n}}. Where n is a number.
+ *  Host replies back with {{echo;m*n}}. Where m is a multiplier both target
+ *  and host know based on the target state.
+ *
+ *        state     multiplier
+ *          0           1
+ *          1           2
+ *
+ *  Response is asserted to verify the state based execution of both host
+ *  and target before and after reset.
+ */
+void do_echo_test(int response_multiplier)
+{
+    char key[] = "echo";
+    char data[] = "10";
+
+    for (int i =0; i < 10; i++)
+    {
+        snprintf(data, sizeof(data), "%d", i);
+        greentea_send_kv("echo", data);
+
+        TEST_ASSERT_NOT_EQUAL_MESSAGE(0, greentea_parse_kv(key, data, sizeof(key), sizeof(data)), "MBED: Failed to recv/parse key \"state\" value from host test!");
```

Comment:
```c++
greentea_parse_kv()
```
id blocking and if ends always returns ```1```. It will return ```0``` ifnly if end of stream was found which will never happen if we read from serial port. E.g. it can return ```0``` when we get ```EOF``` from ```getchar();```.

```c++
greentea_parse_kv(key, data, sizeof(key), sizeof(data));
TEST_ASSERT_EQUAL_STRING("echo", key);
int count = atoi(data);  // i and response_multiplier are both int
TEST_ASSERT_EQUAL_MESSAGE(i * response_multiplier, count, "MBED: Failed to verify count received in echo!");
```